### PR TITLE
chore(guidelines): update the section on error messages

### PR DESCRIPTION
**Pull request by rix0rrr**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by njlynch — APPROVED** [review]

File: DESIGN_GUIDELINES.md [hunk 1]
```diff
@@ -1237,18 +1224,55 @@ example, VPC has **enableDnsHostnames** and **enableDnsSupport**. DNS hostnames
...
+which the user should be alerted to and fix as quickly as possible. Choose attaching
+an error to a construct if the failure is due to environmental factors outside the
+direct use of the API surface (for example, lack of context provider lookup values).
+
+#### Throwing exceptions
+
+This should be the preferred error reporting method.
+
+Validate input as early as it is passed into your code (ctor, methods,
+etc) and bail out by throwing an **Error**. No need to create subclasses of
```
Comment (eladb) [inline DESIGN_GUIDELINES.md:1256]:
```suggestion
etc) and bail out by throwing an `Error`. No need to create subclasses of
```

File: DESIGN_GUIDELINES.md [hunk 2]
```diff
@@ -1237,18 +1224,55 @@ example, VPC has **enableDnsHostnames** and **enableDnsSupport**. DNS hostnames
...
+Error since all errors in the CDK are unrecoverable.
+
```
Comment (eladb) [inline DESIGN_GUIDELINES.md:1258]: Example?

File: DESIGN_GUIDELINES.md [hunk 3]
```diff
@@ -1237,18 +1224,55 @@ example, VPC has **enableDnsHostnames** and **enableDnsSupport**. DNS hostnames
...
 error, this should be modeled in a return value and not through exceptions.
 
-#### Post Validation
+#### Attaching (lazy) Validators
 
-In the rare case where the integrity of your construct can only be checked right
-before synthesis, override the **Construct.validate()** method and return
-meaningful errors. Always prefer early input validation over post-validation.
+In the rare case where the integrity of your construct can only be checked
+right before synthesis, call the **this.node.addValidation()** method to add
```
Comment (eladb) [inline DESIGN_GUIDELINES.md:1267]:
```suggestion
only after the app had completed its initialization, call the **this.node.addValidation()** method to add
```

File: DESIGN_GUIDELINES.md [hunk 4]
```diff
@@ -1237,18 +1224,55 @@ example, VPC has **enableDnsHostnames** and **enableDnsSupport**. DNS hostnames
...
+a validation object. This will generally only be necessary if you want to
+produce an error when a certain interaction with your construct--for example,
+a method call--did *not* happen.
 
-#### Attached Errors/Warnings
+Always prefer early input validation over post-validation, as the necessity
+of these should be rare.
+
```
Comment (eladb) [inline DESIGN_GUIDELINES.md:1274]: Example

**Review by eladb — APPROVED** [review]

File: DESIGN_GUIDELINES.md [hunk 5]
```diff
@@ -1237,18 +1224,55 @@ example, VPC has **enableDnsHostnames** and **enableDnsSupport**. DNS hostnames
 *require* DNS support, so only fail if the user enabled DNS hostnames but
 explicitly disabled DNS support. Otherwise, auto-enable DNS support for them.
 
+#### Error reporting mechanism
+
+There are three mechanism you can use to report errors:
+
+* Eagerly throw an exception (fails synthesis)
+* Attach a (lazy) validator to a construct (fails synthesis)
+* Attach errors to a construct (fails deployment)
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply DESIGN_GUIDELINES.md:1233]: Which error are you talking about?

And if synthesis fails, deployment cannot even proceed so I though it was pointless to put it here. But granted, what's written here might not be clear enough if you're not too intimately familiar with the lifecycle. I was trying to be terse but I may be erroring...

[the remark this answers is not in the document]
Comment (eladb) [reply DESIGN_GUIDELINES.md]: I am talking about errors that are attached to constructs. Those errors will fail the CLI synth (and deploy of course) but will not fail the app initialization and synthesis.

**Pull request merged**